[MIG][13.0] partner_ref_unique: Migration to v13.0

joao-p-marques opened the pull request:
cc @pedrobaeza @Yajo 
TT25339

antespi committed:
partner_ref_unique addon and change description icon

yajo committed:
Work with contact deduplicator.

This contraint is self-disabled when it detects by context that it is being used from the contact deduplicator.

Otherwise, the user cannot deduplicate a contact selecting the "good" one as the one without reference.

Anyway, in these cases, the check has no meaning, since it is to avoid duplicates and in this context we are currently removing them.

cubells committed:
[9.0][ADD] partner_ref_unique module

- Added OCA icon

oca-transbot committed:
OCA Transbot updated translations from Transifex

OCA Transbot updated translations from Transifex

OCA Transbot updated translations from Transifex

OCA Transbot updated translations from Transifex

OCA Transbot updated translations from Transifex

OCA Transbot updated translations from Transifex

cristinamartinrod committed:
[11.0][MIG] partner_ref_unique: Migrate to v11

pedrobaeza committed:
[IMP] partner_ref_unique: Extend tests use cases

OCA-git-bot committed:
[UPD] README.rst

oca-travis committed:
[UPD] Update partner_ref_unique.pot

sbejaoui committed:
[MIG] - migration partner_ref_unique

OCA-git-bot committed:
[UPD] README.rst

oca-travis committed:
[UPD] Update partner_ref_unique.pot

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: partner-contact-12.0/partner-contact-12.0-partner_ref_unique
Translate-URL: https://translation.odoo-community.org/projects/partner-contact-12-0/partner-contact-12-0-partner_ref_unique/

max3903 committed:
[FIX] partner_ref_unique: Pylint errors

OCA-git-bot committed:
[UPD] README.rst

badbole committed:
Translated using Weblate (Croatian)

Currently translated at 100.0% (8 of 8 strings)

Translation: partner-contact-12.0/partner-contact-12.0-partner_ref_unique
Translate-URL: https://translation.odoo-community.org/projects/partner-contact-12-0/partner-contact-12-0-partner_ref_unique/hr/

grindtildeath committed:
[FIX] partner_ref_unique: Trigger check on company update

Before this commit, the partner_ref_unique could be changed on the
company even if existing partners share the same internal ref.

Therefore, errors will appear afterwards as DB already contains
duplicates although the policy shouldn't allow it.

With this change, the error will appear as soon as the policy is
changed, so that the data must be fixed before activating the
constraint.

grindtildeath committed:
Store partner_ref_unique to allow triggering the error updating res.company

Add pre_init_hook to speed up module installation

grindtildeath committed:
fixup! Store partner_ref_unique to allow triggering the error updating res.company

oca-travis committed:
[UPD] Update partner_ref_unique.pot

OCA-git-bot committed:
partner_ref_unique 12.0.1.0.2

oca-transbot committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: partner-contact-12.0/partner-contact-12.0-partner_ref_unique
Translate-URL: https://translation.odoo-community.org/projects/partner-contact-12-0/partner-contact-12-0-partner_ref_unique/

Commit:
Translated using Weblate (Spanish)

Currently translated at 100.0% (8 of 8 strings)

Translation: partner-contact-12.0/partner-contact-12.0-partner_ref_unique
Translate-URL: https://translation.odoo-community.org/projects/partner-contact-12-0/partner-contact-12-0-partner_ref_unique/es/

psc committed:
Translated using Weblate (German)

Currently translated at 100.0% (8 of 8 strings)

Translation: partner-contact-12.0/partner-contact-12.0-partner_ref_unique
Translate-URL: https://translation.odoo-community.org/projects/partner-contact-12-0/partner-contact-12-0-partner_ref_unique/de/

joao-p-marques committed:
[IMP] partner_ref_unique: black, isort, prettier

joao-p-marques committed:
[MIG] partner_ref_unique: Migration to 13.0

joao-p-marques commented:
changes made @pedrobaeza

pedrobaeza commented on partner_ref_unique/models/res_partner.py:
```diff
@@ -1,38 +1,29 @@
-                mode == "all" or (mode == "companies" and partner.is_company)
-            ):
+        for partner in self.filtered("ref"):
+            # If the company is not defined, adopt default strategy for user company
```
This comment is outdated, you can update it to:
```suggestion
            # If the company is not defined in the partner, take current user company
```

pedrobaeza commented on partner_ref_unique/tests/test_res_partner_ref.py:
```diff
@@ -9,49 +10,83 @@ class TestResPartnerRefUnique(common.SavepointCase):
-        cls.partner2 = cls.partner_obj.create({"name": "Partner2",})
+        cls.company_obj = cls.env["res.company"]
+        # look for possible already duplicated refs for being resilient
+        groups = cls.partner_obj.read_group([], ["ref"], ["ref"],)
+        for group in groups:
+            if group["ref_count"] > 1 and not group["ref"]:
+                cls.partner_obj.search(group["__domain"]).write({"ref": False})
```
We talked about resetting all refs:
```suggestion
        groups = cls.partner_obj.search([("ref", "!=", False)]).write({"ref": False})
```

pedrobaeza commented on partner_ref_unique/models/res_partner.py:
```diff
@@ -1,38 +1,29 @@
                     domain.append(("is_company", "=", True))
                 other = self.search(domain)
-
-                # active_test is False when called from
```
We haven't analyzed this. @Yajo do you remember why this dirty trick for merging. What is the problem merging?

yajo replied:
I don't remember, but it makes sense that one wants to merge archived partners too.

I think this code is just trying to avoid the validation if you're merging, maybe because during the merge there'll be a valid situation where 2 partners will share the same ref always? :thinking: 

This context doesn't seem to be the most reliable way of doing that, OTOH.

pedrobaeza replied:
That context key was a trick for detecting that you are coming "from a merge operation", so indeed is not correct to check that (you can inject one context key if needed). @joao-p-marques you can check to put one partner with ref, and another without, activate the check, and try to merge one into another (try both ways) to see if there's such problem.

pedrobaeza replied:
Indeed there's a problem when merging that we have already handled. See 9a8c078

pedrobaeza commented on partner_ref_unique/tests/test_res_partner_ref.py:
```diff
@@ -9,49 +10,83 @@ class TestResPartnerRefUnique(common.SavepointCase):
         with self.assertRaises(ValidationError):
             self.partner_obj.create(
-                {"is_company": True, "name": "other", "ref": "same_ref",}
+                {"name": "other", "ref": "same_ref", "is_company": True}
```
You can let this line without changes

pedrobaeza commented on partner_ref_unique/tests/test_res_partner_ref.py:
```diff
@@ -9,49 +10,83 @@ class TestResPartnerRefUnique(common.SavepointCase):
+                {"name": "other", "ref": "same_ref", "is_company": True}
             )
-        # Here there shouldn't be any problem
-        self.partner_obj.create(
```
This one is useful to be let as well.

pedrobaeza reviewed (approved):
Waiting the merge partner confirmation

yajo commented on partner_ref_unique/models/res_company.py:
```diff
@@ -16,3 +17,17 @@ class ResCompany(models.Model):
+            partners = (
+                self.env["res.partner"]
+                .with_context(active_test=False)
+                .search([("company_id", "in", [False] + self.ids)])
```
I think this would be better.

```suggestion
                .search([("company_id", "in", self.ids)])
```

Otherwise different companies with different settings would collide one another.

pedrobaeza replied:
No, it's correct as it is, as we want to specifically check both. Note that this is not executed with `sudo`, so the search is performed on the user environment (global contacts + current company contacts).

yajo replied:
Are you sure? AFAICS what could happen is that the user logs as company A, sets one policy, company-less contacts pass; then logs as company B, sets a different policy, and company-less contacts fail. :roll_eyes:

pedrobaeza replied:
I don't get you, but yes, I'm sure. If you don't, please expose detailed steps to reproduce a problem.

yajo replied:
On my way, just hold on.

yajo replied:
Yes, this creates an inconsistent UX:

1. Create Company A, with uniqueness check for all partners; create company B, with None: 
![crear compañías](https://user-images.githubusercontent.com/973709/92569664-d0205a80-f278-11ea-8f4b-f23d9308adf4.gif)
1. Go to contacts, switch to company B, create 2 contacts without company that have the same reference: 
![sameref](https://user-images.githubusercontent.com/973709/92569814-f9d98180-f278-11ea-9d5c-ed648e6d32aa.gif)
   No problem because company B is allowed to do that.
1. Switch to company A, go to companies, edit company A, remove uniqueness constraint, add it again. Error: 
![error](https://user-images.githubusercontent.com/973709/92570167-6ce2f800-f279-11ea-9094-ef8501b2003e.gif)

So, users are able to duplicate refs of partners that have no company if they are in the context of company B, but on under company A. Effectively, this means that the constraint is inconsistent, thus not very useful for companyless partners.

A solution would be to have an ICP for those partners that is consistently enforced across companies, or another solution would be that companyless partners inherit the most strict convention of all present companies.

pedrobaeza replied:
We already thought about ICP, but it's more convoluted. The case you depict can happen the same with the ICP if the combination is not good, but it's not directly caused by including here the `False` part, and not including it will cause a bunch more of incorrect cases, so let's continue with this as is. Think that previously the duplications were worse, and it always force to have a company in the partners to work.

pedrobaeza commented on partner_ref_unique/hooks.py:
```diff
@@ -0,0 +1,33 @@
+# Copyright 2020 Camptocamp SA
```
Oh, we forget this! This can be fully removed and the hook removed in the manifest.

joao-p-marques replied:
Yes, thank you. Changes already made and pushed.

HaraldPanten commented on partner_ref_unique/README.rst:
```diff
@@ -0,0 +1,102 @@
+
+|badge1| |badge2| |badge3| |badge4| |badge5| 
+
+This module allows add an unique constraint to partner ref field.
```
```suggestion
This module allows add a unique constraint to partner ref field.
```

HaraldPanten commented on partner_ref_unique/README.rst:
```diff
@@ -0,0 +1,102 @@
+#. Go to any partner in **Contacts > Contacts**
+#. Try to create two partners with the same ref.
+
+* If you selcted the option 'All partners' you can't create two partners with the same ref.
+* If you selcted the option 'Only companies' you can't create two companies with the same ref.
+* If you selcted the option 'None' you can create two partners with the same ref.
```
```suggestion
* If you selected the option 'All partners' you can't create two partners with the same ref.
* If you selected the option 'Only companies' you can't create two companies with the same ref.
* If you selected the option 'None' you can create two partners with the same ref.
```

Typo

pedrobaeza replied:
It's not correct this way either:
```
* If you select the option 'None', you can...
```

pedrobaeza replied:
It's correct as is, sorry.

HaraldPanten commented on partner_ref_unique/README.rst:
```diff
@@ -0,0 +1,102 @@
+
+To configure this module, you need to:
+
+#. Go to **Settings > Companies**.
```
```suggestion
#. Go to **Settings > Users & Companies > Companies**.
```

HaraldPanten reviewed (approved):
Functional review. Just some minor picky comments 😝👍

joao-p-marques commented:
Functional changes made as suggested by @HaraldPanten . Thank you!

pedrobaeza committed:
[IMP] partner_ref_unique: Take into account contact merging

If merging 2 contacts, one with reference, and another without, and being the target
one the contact without reference, the constraint is raised as the reference value
is copied before unlinking the other contact.

We avoid it disabling the constraint when merging.

pedrobaeza commented:
/ocabot merge nobump

pedrobaeza commented:
/ocabot merge nobump
